integration/experiment/system_characterization/sys_charac_utils.sh

```diff
@@ -0,0 +1,111 @@
+#!/bin/bash
+
+set -e
+set -x
+
+module load geopm-runtime
+
+SWEEP_TYPE="${SWEEP_TYPE?-Set SWEEP_TYPE to either \"CPU\" or \"GPU\"}"
+PROGRAM_NAME="${PROGRAM_NAME?-Set PROGRAM_NAME to the binary name of the characterization application}"
+BINARY_PATH_PLUS_FLAGS="${BINARY_PATH_PLUS_FLAGS?-Set BINARY_PATH_PLUS_FLAGS to the full path of the application binary including the arguments}"
+EMPTY_SWEEP_OUTPUT_DIR="${EMPTY_SWEEP_OUTPUT_DIR?-Set EMPTY_SWEEP_OUTPUT_DIR to the output directory of the characterization sweeps}"
+INIT_CONTROLS_LIST="${EMPTY_SWEEP_OUTPUT_DIR}/init_controls-${HOSTNAME}.lst"
+
+TRIAL_COUNT=3
+
+CPU_LOGICAL_CORE_CNT=$(geopmread -d | grep cpu | awk '{print $2}')
+CPU_PHYSICAL_CORE_CNT=$(geopmread -d | grep core | awk '{print $2}')
+CPU_LOGICAL_CORE_CNT_MINUS_1=$(( ${CPU_LOGICAL_CORE_CNT} - 1 ))
+CPU_PHYSICAL_CORE_CNT_MINUS_4=$(( ${CPU_PHYSICAL_CORE_CNT} - 4 ))
+
+mkdir -p $EMPTY_SWEEP_OUTPUT_DIR
+echo "Writing logs and reports to ${EMPTY_SWEEP_OUTPUT_DIR}"
+
+launch_sweep () {
+
+          geopmlaunch pals \
+            -n ${RANK_COUNT} -ppn ${RANK_COUNT} \
+            --cpu-bind list:0-${CPU_LOGICAL_CORE_CNT_MINUS_1} \
+            --geopm-init-control=$INIT_CONTROLS_LIST \
+            --geopm-ctl=application \
+            --geopm-preload \
+            --geopm-profile="${PROGRAM_NAME}" \
+            --geopm-report=${REPORT_FILE_PATH} \
+            --geopm-report-signals=${GEOPM_SIGNALS} \
+            --geopm-program-filter=${PROGRAM_NAME} \
+            -- ${BINARY_PATH_PLUS_FLAGS} 2>&1 \
+            > ${LOG_FILE_PATH}
+
+          sleep 5
+}
+
+
+launch_cpu_sweep () {
+
+                RANK_COUNT=${CPU_PHYSICAL_CORE_CNT_MINUS_4}
```

Comment:
Include some comments about the chosen rank count per node:
1. How you arrived at this configuration.
2. Why it's appropriate for this type of sweep.